zebra-chain/src/types/amount.rs

```diff
@@ -9,6 +9,102 @@ use std::{
 #[derive(Debug, Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
 pub struct Amount<C = NegativeAllowed>(i64, PhantomData<C>);
 
+impl<C1, C2> std::ops::Add<Amount<C2>> for Amount<C1>
+where
+    C1: AmountConstraint,
+    C2: AmountConstraint,
+{
+    type Output = Option<Amount<C1>>;
```

Comment:
Hmm, if `C1` and `C2` are different constraints, how do we combine them into a `C1` for the `Output`?

Should this maybe be
```
impl<C: AmountConstraint> Add<Amount<C>> for Amount<C> {
    type Output = Option<Amount<C>>
    // ...
}
```

Reply:
So this combines the constraints by doing the addition outside of the `Amount`. It would just unwrap the two Amounts regardless of their constraints and then it picks one of the constraints as the output one and converts the result of the op into an Amount of that constraint, validating it along the way. I decided to default to the LHS as the constraint that gets used for the result because it seemed more intuitive. This lets you do things like `positive_amount + negative_amount` and be sure that the result is still positive.

If you think we should never add amounts with different constraints then we will probably need to add some conversion helpers to convert amounts of one constraint type into amounts of another constraint, however we probably can't do this with `From` because it will almost certainly overlap with `From<T> for T`.

Reply:
Yeah, what I mean is combining the constraints like you mention.  I think we should never add amounts with different constraints, unless we have a principled way to determine what the output constraint should be.  Otherwise, computing
```
positive_amt + negative_amt
```
and
```
negative_amt + positive_amt
```
would have different answers even though addition is supposed to be commutative. I think that this would be really surprising and error-prone.

We should definitely have a way to convert constraints as you mention.  If coherence prevents us from writing
```
impl<C1, C2> TryFrom<Amount<C1>> for Amount<C2>
```
then my next-best idea would be to write
```
impl<C1> Amount<C1> {
    pub fn constrain<C2>(&self) -> Result<Amount<C2>>
}
```
so that we can do
```
let positive_amount = negative_amount.constrain::<NonNegative>();
```

Reply:
We might be able to add a 3rd parameter for the output and let inference take over.

Reply:
> If coherence prevents us from writing

Actually, thinking about it, coherence only disallows _generic_ impls for this, we can just manually specify the conversions between the two constraints we actually have I think.

Reply:
I tried both using a manual `From`/`TryFrom` impls and ended up abandoning it because it wasn't able to infer the correct type for `into` through the generics on `Add`, the suggested `constrain` API works with inference quite nicely though so I went ahead and added it.